llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp

```diff
@@ -308,31 +375,97 @@ void transform_ext::makeReductionMatcher(
+          .dim(-3, utils::IteratorType::parallel, /*strictInBounds=*/false)
+          .dim(-4, utils::IteratorType::parallel, /*strictInBounds=*/false)
+          // Single input for now, can be arbitrary projected permutations.
+          // TODO: Multiple inputs, can be arbitrary projected permutations.
+          // TODO: Watch out for multiple inputs though as a reduction turns
+          //       into a contraction when mixed with projected
+          //       permutations. A reduction is often bandwidth bound but
+          //       contraction is a different beast that is compute bound
+          //       and has a very different schedule.
+          //
+          .input(NumEqualsTo(1))
+          .input(AllOperands(), IsPermutation().setProjectedPermutation(true))
+          // Single output supported atm.
+          // TODO: Multiple outputs.
+          //
+          .output(NumEqualsTo(1))
+          // A reduction output must be a projected permutation, match it but we
+          // could also drop this technically.
+          .output(AllOperands(), IsPermutation().setProjectedPermutation(true))
+          // Only single combiner over 32 bits for now due to reduction warp
+          // distribution.
+          // TODO: relax this once reduction distribution is more powerful.
+          //
+          .output(0, ElementTypeBitWidth(32))
+          .output(0, SingleCombinerReduction());
+
+  // Mandatory FillOp must create the unique output of the reduction.
+  // TODO: Relax this, as any map, broadcast, transpose should also work.
+  //
   fill = m_StructuredOp<linalg::FillOp>();
-  trailing = m_StructuredOp<linalg::GenericOp>()
-                 .input(AllOperands(), IsPermutation())
-                 .output(AllOperands(), IsPermutation())
-                 .input(NumEqualsTo(1))
-                 .output(NumEqualsTo(1));
-  leading = trailing;
-  reduction = m_StructuredOp()
-                  // The CUDA strategy now supports arbitray mixes of static and
-                  // dynamic sizes and adapts accordingly.
-                  // Consider separating control for other targets if needed.
-                  .dim(-1, utils::IteratorType::reduction)
-                  .dim(-1, CaptureDim(reductionDimensionSize))
-                  // Can be extended to projected permutation with broadcast.
-                  .input(AllOperands(), IsPermutation())
-                  // TODO: we want to accept any input position here.
-                  .input(0, leading, OptionalMatch())
-                  .output(NumEqualsTo(1))
-                  .output(0, fill)
-                  // Only single combiner over 32 bits for now due to
-                  // reduction distribution.
-                  .output(0, ElementTypeBitWidth(32))
-                  .output(0, SingleCombinerReduction())
-                  .result(0, HasAnyUse(), trailing, OptionalMatch());
+  reduction = reduction.output(NumEqualsTo(1)).output(0, fill);
```

Comment:
I'm not convinced by this decomposition. It makes it easy to miss what actually goes into the matcher. I'd rather keep one long chain.